TASK: Raise PhpStan to level 8 - Part 1

mficzel opened the pull request:
Raise PhpStan to level 8 for: 

- `Neos.Flow.Log`
- `Neos.FluidAdaptor`
- `Neos.Http.Factories`
- `Neos.Kickstarter`
- `Neos.Utility.Arrays`
- `Neos.Utility.Files`
- `Neos.Utility.MediaTypes`
- `Neos.Utility.ObjectHandling`
- `Neos.Utility.OpcodeCache`
- `Neos.Utility.Pdo`
- `Neos.Utility.Schema`
- `Neos.Utility.Unicode`

**Upgrade instructions**

No changes needed.

**Review instructions**

This the first of multiple PRs to cherry pick the changes from https://github.com/neos/flow-development-collection/pull/3474 from @nezaniel and @mhsdesign step by step and fix issues and breakiness. 

This is intended to be unbreaky but be the foundation for future refactorings that can at least partly be automated once we have really solid type informations. 

**Checklist**

- [x] Code follows the PSR-2 coding style
- [x] Tests have been created, run and adjusted as needed
- [x] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

Commit:
WIP: Adjust Neos.Utility.Unicode

Commit:
Adjust Neos.Utility.Schema

Commit:
Adjust Neos.Utility.Pdo

Commit:
Adjust Neos.Utility.ObjectHandling

Commit:
Adjust Neos.Utility.MediaTypes

Commit:
Adjust Neos.Utility.Files

Commit:
Adjust Neos.Utility.Arrays

Commit:
Adjust Neos.Kickstarter

Commit:
Adjust Neos.Http.Factories

Commit:
Adjust Neos.Http.Factories

Commit:
Adjust Neos.FluidAdaptor

Commit:
Adjust Neos.Flow.Log

Commit:
Adjust Neos.Http.Factories to Neos.Flow Http subcontext adjustments

Commit:
add scanDirectories folders for special classes

mficzel committed:
TASK: Revert breaking changes to Utility->ObjectAccess

mficzel committed:
TASK: Revert breaking changes to Utility->Arrays

mficzel committed:
TASK: Adjust phpstan config to apply level 8 rules to the utility packages and use level 3 otherwise

mficzel committed:
TASK: Adjust utility packages to phpstan level 8

mficzel committed:
TASK: Revert breaking changes of Utility.MediaTypes

mficzel committed:
TASK: Adjust Http.Factories and Kickstarter for Stan level 8

mficzel committed:
TASK: Adjust FluidAdapter for PhpStan level 8

mficzel committed:
TASK: Raise PhpStan level for Flow.Log to 8 and adjust breakiness

mhsdesign commented:
wow thanks a lot! If you need help here let me know;) Also lets ignore the style CI failures as we want to move to php cs either way and they might have different expectations again;)

mficzel commented:
@mhsdesign is suggest to merge this in parts as otherwise the reviews will get really overwhelming.

mhsdesign commented:
oke i assume my commits will be part of another change then? idk why i did those but i probably did them to get flow back running

<img width="710" height="337" alt="image" src="https://github.com/user-attachments/assets/b16b3e80-5bef-44ad-b8c4-7a633094b05a" />

mficzel commented:
@mhsdesign yes your commits were mostly regarding flow. I cherry pick commits for each section one by one and afterwards remove breakiness and make tests green + verify neos still works.

mhsdesign commented:
Okay so this is the final set that should be tested and go into Flow 9.1 then ill review and test that ;) Thanks for splitting this mammoth!

mficzel commented:
I will try to prepare follow ups that may or may not be ready for 9.1 but imho this is about the maximal size for a pr that is chewable. So imho this can be reviewed and merged

mhsdesign committed:
TASK: Add phpstan template/generics to `getAllSubClassNamesForClass()`

mhsdesign committed:
PATCH: Dont annotate `DependencyProxy` in `@var` as this is an antipattern in flow

if this would be injected flow would not handle this

mhsdesign committed:
PATCH: Revert to correct types in `PaginateController`

the change in c6e32474c0fa0ce1940c2d2e6fb9aa7f87f3d566 was half thought and phpstan only complained due to temporary results written.

mhsdesign committed:
PATCH: Remove obsolete `@var` type hint (`isMainRequest` already tells phpstan whats up)

mhsdesign committed:
PATCH: Document what _legacy_ phpstan means (lvl3) and fix one whitespace *g

mhsdesign committed:
BUGFIX: Throw exception when `Files::getRecursiveDirectoryGenerator()` is used with `$returnRealPath` but realpaths cannot be calculated

This is the case when using a virtual file system for testing in example (`vfs://`) or really any stream wrapper.

Silently swallowing all the `realpath` misses disguises bugs and hiders developer experience when using a vfs.

mhsdesign reviewed (approved):
Thanks for reviving this change and battling with it to make it run :D

I did a full review locally and pushed some small adjustment changes - just mostly syntactical stuff.

I see we now use a lot of times the pattern `?: ''` to convert a `false` to a string. That is often okay as this the `false` case should never happen for constant preg expressions but in some cases ignoring this type juggling disguises bugs. So if there is real reason to believe it _could_ happen im definitely for having a sane fallback or telling whats wrong.

Thats exactly why i added a new exception to `\Neos\Utility\Files::getRecursiveDirectoryGenerator` see https://github.com/neos/flow-development-collection/pull/3515/commits/5769192c81c8450c10f9ed2c1744f8d343343e82:

> BUGFIX: Throw exception when `Files::getRecursiveDirectoryGenerator()` is used with `$returnRealPath` but realpaths cannot be calculated
>
> This is the case when using a virtual file system for testing in example (`vfs://`) or really any stream wrapper.
>
> Silently swallowing all the `realpath` misses disguises bugs and hiders developer experience when using a vfs.

mhsdesign committed:
Merge remote-tracking branch 'origin/9.1' into phpstanLevel8-v3

mhsdesign committed:
TASK: Throw if `rotateLogFile` cannot create or open lock file

`fopen` is multi thread safe. $lockResource is never `false` unless something is severely wrong with the system like flow has no write access to Data/
